src/actions/libraries/AssemblyUtils.sol

```diff
@@ -3,25 +3,68 @@ pragma solidity ^0.8.26;
     function safeTransferNoReturn(address target, uint256 value, uint256 gasLimit) internal returns (bool success) {
         assembly {
-            // Call the target address with the specified value and gas limit
-            // 1. Set the gas limit for the call
-            // 2. Set the target to send ether to
-            // 3. Set the value to send (in wei)
-            // 4. Set the call data pointer to zero (no calldata needed)
-            // 5. Set the call data size to zero (no calldata needed)
-            // 6. Set the return data pointer to zero (discard return data)
-            // 7. Set the return data size to zero (discard return data)
+            // call the target address with the specified value and gas limit
+            // 1. set the gas limit for the call
+            // 2. set the target to send ether to
+            // 3. set the value to send (in wei)
+            // 4. set the call data pointer to zero (no calldata needed)
+            // 5. set the call data size to zero (no calldata needed)
+            // 6. set the return data pointer to zero (discard return data)
+            // 7. set the return data size to zero (discard return data)
             success := call(gasLimit, target, value, 0, 0, 0, 0)
         }
     }
+
+    /// @notice safely calls a target contract with a gas limit and return data size limit
+    /// @dev this utility avoids return bomb attacks by limiting the size of the return data
+    /// @param target - the address to call
+    /// @param value - the amount of ether to send (in wei)
+    /// @param gasLimit - the maximum amount of gas to use for the call
+    /// @param returnSizeLimit - the maximum size of the return data (in bytes)
+    /// @param data - the call data to send to the target contract
+    /// @return success - whether the transfer was successful or not
+    /// @return result - the return data from the call, limited to `returnLimit` bytes
+    function safeCallLimitReturn(
+        address target,
+        uint256 value,
+        uint256 gasLimit,
+        uint16 returnSizeLimit,
+        bytes memory data
+    ) internal returns (bool success, bytes memory result) {
+        // allocate `returnSizeLimit` bytes of memory for the return data
+        result = new bytes(returnSizeLimit);
+
+        assembly {
+            // call the target address with the specified value and gas limit
+            // 1. set the gas limit for the call
+            // 2. set the target to send ether to
+            // 3. set the value to send (in wei)
+            // 4. set the call data pointer to start of `data` buffer
+            // 5. set the call data size to the length of `data`
+            // 6. set the return data pointer to zero (manually copy return data)
+            // 7. set the return data size to zero (manually copy return data)
+            success := call(gasLimit, target, value, add(data, 0x20), mload(data), 0, 0)
```

Comment:
why `add(0x20)`?

Reply:
Under the hood: the first 32 bytes of memory buffers are reserved for the length of the buffer for quick access, so here we're saying "skip over the first 32 bytes and set the calldata pointer to where the actual data starts". Similarly for the next parameter (calldata length), we load the first 32 bytes at the `data` pointer which stores the length of the calldata